src/Middleware/RequestThrottling/src/Internal/RequestQueue.cs

```diff
@@ -9,36 +9,40 @@ namespace Microsoft.AspNetCore.RequestThrottling.Internal
 {
     internal class RequestQueue : IDisposable
     {
-        private SemaphoreSlim _semaphore;
+        private readonly int _maxConcurrentRequests;
+        private readonly int _requestQueueLimit;
+        private readonly SemaphoreSlim _semaphore;
+
         private object _waitingRequestsLock = new object();
-        public readonly int MaxConcurrentRequests;
-        public int WaitingRequests { get; private set; }
+        private int _waitingRequests;
 
-        public RequestQueue(int maxConcurrentRequests)
+        public RequestQueue(int maxConcurrentRequests, int requestQueueLimit)
         {
-            MaxConcurrentRequests = maxConcurrentRequests;
+            _maxConcurrentRequests = maxConcurrentRequests;
+            _requestQueueLimit = requestQueueLimit;
             _semaphore = new SemaphoreSlim(maxConcurrentRequests);
         }
 
-        public async Task EnterQueue()
+        public async Task<bool> TryEnterQueueAsync()
         {
-            var waitInQueueTask = _semaphore.WaitAsync();
-
-            var needsToWaitOnQueue = !waitInQueueTask.IsCompletedSuccessfully;
-            if (needsToWaitOnQueue)
+            lock (_waitingRequestsLock)
             {
-                lock (_waitingRequestsLock)
+                if (_waitingRequests >= _requestQueueLimit)
                 {
-                    WaitingRequests++;
+                    return false;
                 }
 
-                await waitInQueueTask;
+                _waitingRequests++;
+            }
+
+            await _semaphore.WaitAsync();
```

Comment:
One optimization I'd be fine with you doing now (or saving in your back pocket for when you start benchmarking) is to store the Task returned by WaitAsync. If it's already complete, you can avoid taking the _waitingRequestsLock once let alone twice, since you wouldn't need to increment or decrement _waitingRequests.